### DISCUSSION TITLE
How to make sum(w)=0 and sum(abs(x))=1 when max sharpe ratio?

### DISCUSSION THREAD

**Discussion by temph2020** [skfolio/skfolio#84] [General]
when maximizing sharpe ratio, I want sum(w)=0 and sum(abs(x))=1, how to do it?
I tried the follollowing code, but get error:
    raise DCPError(
cvxpy.error.DCPError: Problem does not follow DCP rules. Specifically:
The following constraints are not DCP:
0.5 <= Sum(maximum(var1, 0.0), None, False) , because the following subexpressions are not:
|--  0.5 <= Sum(maximum(var1, 0.0), None, False)
0.5 <= Sum(-minimum(var1, 0.0), None, False) , because the following subexpressions are not:
|--  0.5 <= Sum(-minimum(var1, 0.0), None, False)

###########################################################
I tried the following code:
prices = load_sp500_dataset()

X = prices_to_returns(prices)
X_train, X_test = train_test_split(X, test_size=0.33, shuffle=False)

def custom_constraints(w): 
    return [cp.sum(cp.pos(w))>=0.5,cp.sum(cp.neg(w))>=0.5]

model = MeanRisk(
    risk_measure=RiskMeasure.VARIANCE,
    objective_function=ObjectiveFunction.MAXIMIZE_RATIO,
    portfolio_params=dict(name="Max Sharpe"),
    budget = 0, # market neutral portfolio.
    min_weights= -1,
    max_short = 0.5,
    max_long= 0.5,
    add_constraints=custom_constraints
    
)
model.fit(X_train)
print(model.weights_)
print('sum weight',sum(model.weights_),'sumabs',sum([abs(w) for w in model.weights_]))

**Comment by HugoDelatte** [comment]
Hello,  
Minimizing a convex function (e.g., variance) under the equality constraint sum(abs(x)) = 1 is a non-convex problem. Although abs(x)  is convex, an equality constraint on a convex function generally makes the problem non-convex. This is why convex solvers cannot solve it and why CVXPY flags it as not compliant with the DCP rules, which require equality constraints to be linear.

**Comment by temph2020** [comment]
Yes, I know it's non-convex. I just wonder if there exists a transformation which could reformulate it to a convex problem.

**Comment by HugoDelatte** [comment]
Not to my knowledge. 
Typically, the issue with market-neutral portfolio optimization without additional constraints is that all your weights could end up being all zeros. Since an equality constraint on absolute values cannot be used, practitioners sometimes use group constraints to avoid this problem. For example, the sum of all bond weights >=+50%. 

Also, even if such constraints may seem too restrictive, in practice, they tend to be beneficial due to their regularization effect (reducing overfitting and out-of-sample stability, especially in long-short portfolios).

**Comment by temph2020** [comment]
Thanks. I do not use bonds. Only stocks, the universe is 1000 stocks. I want to build a long-short market-neutral portfolio with sum(abs(w))=1. Could we use group constraints to avoid this problem?
